contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1427 @@
+
+        uint256 nuStakeAmount = nucypherStakingContract.getAllTokens(
+            operator.owner
+        );
+        (uint96 tAmount, ) = nuToT(nuStakeAmount);
+
+        require(operator.nuInTStake > tAmount, "There is no discrepancy");
+        operator.nuInTStake = tAmount;
+        seizeNu(
+            operator,
+            stakeDiscrepancyPenalty,
+            stakeDiscrepancyRewardMultiplier
+        );
+
+        emit TokensSeized(_operator, tAmount - operator.nuInTStake);
+        authorizationDecrease(_operator, operator);
+    }
+
+    /// @notice Sets the penalty amount for stake discrepancy and reward
+    ///         multiplier for reporting it. The penalty is seized from the
+    ///         operator account, and 5% of the penalty, scaled by the
+    ///         multiplier, is given to the notifier. The rest of the tokens are
+    ///         burned. Can only be called by the Governance. See `seize` function.
+    function setStakeDiscrepancyPenalty(
+        uint96 penalty,
+        uint256 rewardMultiplier
+    ) external override onlyGovernance {
+        // TODO optimization: can save NU and KEEP equivalents to exclude conversion during slashing
+        stakeDiscrepancyPenalty = penalty;
+        stakeDiscrepancyRewardMultiplier = rewardMultiplier;
+        emit StakeDiscrepancyPenaltySet(penalty, rewardMultiplier);
+    }
+
+    /// @notice Sets reward in T tokens for notification of misbehaviour
+    ///         of one operator
+    function setNotificationReward(uint96 reward) external onlyGovernance {
+        notificationReward = reward;
+        emit NotificationRewardSet(reward);
+    }
+
+    /// @notice Transfer some amount of T tokens as reward for notifications
+    ///         of misbehaviour
+    function pushNotificationReward(uint96 reward) external {
+        require(reward > 0, "Reward must be specified");
+        notifiersTreasury += reward;
+        emit NotificationRewardPushed(reward);
+        token.safeTransferFrom(msg.sender, address(this), reward);
+    }
+
+    /// @notice Adds operators to the slashing queue along with the amount that
+    ///         should be slashed from each one of them. Can only be called by
+    ///         application authorized for all operators in the array.
+    function slash(uint96 _amount, address[] memory _operators)
+        external
+        override
+    {
+        notify(_amount, 0, address(0), _operators);
+    }
+
+    /// @notice Adds operators to the slashing queue along with the amount.
+    ///         The notifier will receive reward per each operator from
+    ///         notifiers treasury. Can only be called by application
+    ///         authorized for all operators in the array.
+    function seize(
+        uint96 _amount,
+        uint256 _rewardMultiplier,
+        address _notifier,
+        address[] memory _operators
+    ) external override {
+        notify(_amount, _rewardMultiplier, _notifier, _operators);
```

Comment:
Should we emit an event from slash/seize?

Reply:
I thought about that. But looks like such events will duplicate data from `slashingQueue` without any changes (because elements in `slashingQueue` don't change). I'd prefer to keep it like now

Reply:
Cool, the application can also emit events when calling `slash`/`seize`. Would be good to mention it in `@dev` docs to `slash`/`seize` for future-us.

Reply:
✔️